File: tests/helpers/spec-assert.ts
/**
 * spec-assert.ts — kit-versioned matcher-aware deep-equality comparator (v2.2).
 *
 * The generated `data`/`ui` boundary harness asserts an actual value against
 * `fixtures/<case-id>-expected.json` via `specAssert(actual, expected)`. Deep
 * equality, EXCEPT that volatile fields in the expected fixture carry a matcher
 * token instead of a literal (§5.49). This is the ONLY comparator — no per-project
 * assertion logic. One file per supported test language; this is the TS/JS one.
 *
 * Matcher tokens (string values in the expected fixture):
 *
 *   "<UUID>"            any UUID v4 string
 *   "<ISO8601>"         any ISO 8601 datetime string
 *   "<ANY_STRING>"      any string
 *   "<ANY_NUMBER>"      any finite number
 *   "<UNORDERED>"       any array (presence + shape only)
 *   "<MATCHES:regex>"   any string matching the pattern (RegExp.test)
 *
 * Order-insensitive array WITH item checking: wrap the expected array as
 *   { "<UNORDERED>": [item, item, ...] }
 * and the actual must be an array that is an order-insensitive deep-equal multiset.
 *
 * A token asserts PRESENCE + SHAPE — never "ignore this field". Missing volatile
 * fields fail (strict key sets); extra actual keys also fail.
 *
 *   import { specAssert, loadFixture } from "../helpers/spec-assert";
 *   specAssert(actual, loadFixture("fixtures/PH2-ORDER-01-expected.json"));
 */

import { readFileSync } from "node:fs";

const UUID4 =
  /^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-4[0-9a-fA-F]{3}-[89abAB][0-9a-fA-F]{3}-[0-9a-fA-F]{12}$/;
const ISO8601 =
  /^\d{4}-\d{2}-\d{2}[Tt ]\d{2}:\d{2}:\d{2}(\.\d+)?([Zz]|[+-]\d{2}:?\d{2})?$/;
const MATCHES_PREFIX = "<MATCHES:";

export class SpecMismatch extends Error {
  constructor(message: string) {
    super(message);
    this.name = "SpecMismatch";
  }
}

export function loadFixture<T = unknown>(path: string): T {
  return JSON.parse(readFileSync(path, "utf-8")) as T;
}

function fail(path: string, msg: string): never {
  throw new SpecMismatch(`spec-assert mismatch at ${path}: ${msg}`);
}

function isPlainObject(v: unknown): v is Record<string, unknown> {
  return typeof v === "object" && v !== null && !Array.isArray(v);
}

function typeName(v: unknown): string {
  if (v === null) return "null";
  if (Array.isArray(v)) return "array";
  return typeof v;
}

/** Returns true if `token` is a known matcher and `actual` satisfies it.
 *  Throws on a known token that is NOT satisfied. Returns false if not a token. */
function matchToken(actual: unknown, token: string, path: string): boolean {
  switch (token) {
    case "<UUID>":
      if (!(typeof actual === "string" && UUID4.test(actual)))
        fail(path, `expected a UUID v4 string, got ${JSON.stringify(actual)}`);
      return true;
    case "<ISO8601>":
      if (!(typeof actual === "string" && ISO8601.test(actual)))
        fail(path, `expected an ISO 8601 datetime string, got ${JSON.stringify(actual)}`);
      return true;
    case "<ANY_STRING>":
      if (typeof actual !== "string")
        fail(path, `expected any string, got ${typeName(actual)}`);
      return true;
    case "<ANY_NUMBER>":
      if (typeof actual !== "number" || !Number.isFinite(actual))
        fail(path, `expected any finite number, got ${JSON.stringify(actual)}`);
      return true;
    case "<UNORDERED>":
      if (!Array.isArray(actual))
        fail(path, `expected any array, got ${typeName(actual)}`);
      return true;
  }
  if (token.startsWith(MATCHES_PREFIX) && token.endsWith(">")) {
    const pattern = token.slice(MATCHES_PREFIX.length, -1);
    if (!(typeof actual === "string" && new RegExp(pattern).test(actual)))
      fail(path, `expected a string matching /${pattern}/, got ${JSON.stringify(actual)}`);
    return true;
  }
  return false;
}

function unorderedEqual(actual: unknown, items: unknown[], path: string): void {
  if (!Array.isArray(actual))
    fail(path, `expected an array (unordered), got ${typeName(actual)}`);
  if (actual.length !== items.length)
    fail(path, `array length ${actual.length} != expected ${items.length} (unordered)`);
  const remaining = [...actual];
  items.forEach((exp, i) => {
    let found = -1;
    for (let j = 0; j < remaining.length; j++) {
      try {
        specAssert(remaining[j], exp, `${path}[unordered:${i}]`);
        found = j;
        break;
      } catch (e) {
        if (e instanceof SpecMismatch) continue;
        throw e;
      }
    }
    if (found < 0) fail(path, `no actual element matches expected item #${i}: ${JSON.stringify(exp)}`);
    remaining.splice(found, 1);
  });
}

/** Assert `actual` deep-equals `expected`, honoring matcher tokens. Throws
 *  SpecMismatch with a JSON path on the first divergence. */
export function specAssert(actual: unknown, expected: unknown, path = "$"): void {
  // scalar matcher token
  if (typeof expected === "string" && matchToken(actual, expected, path)) return;

  // { "<UNORDERED>": [...] } wrapper → order-insensitive item compare
  if (isPlainObject(expected)) {
    const keys = Object.keys(expected);
    if (keys.length === 1 && keys[0] === "<UNORDERED>") {
      unorderedEqual(actual, expected["<UNORDERED>"] as unknown[], path);
      return;
    }
    if (!isPlainObject(actual)) fail(path, `expected object, got ${typeName(actual)}`);
    const expKeys = new Set(keys);
    const actKeys = new Set(Object.keys(actual));
    const missing = [...expKeys].filter((k) => !actKeys.has(k));
    const extra = [...actKeys].filter((k) => !expKeys.has(k));
    if (missing.length || extra.length)
      fail(path, `key mismatch (missing=${JSON.stringify(missing)}, extra=${JSON.stringify(extra)})`);
    for (const k of keys) specAssert(actual[k], expected[k], `${path}.${k}`);
    return;
  }

  if (Array.isArray(expected)) {
    if (!Array.isArray(actual)) fail(path, `expected array, got ${typeName(actual)}`);
    if (actual.length !== expected.length)
      fail(path, `array length ${actual.length} != expected ${expected.length}`);
    for (let i = 0; i < expected.length; i++)
      specAssert(actual[i], expected[i], `${path}[${i}]`);
    return;
  }

  // scalar literal — strict equality (typeof must agree; NaN-safe)
  if (typeName(actual) !== typeName(expected))
    fail(path, `type ${typeName(actual)} != expected ${typeName(expected)}`);
  if (actual !== expected && !(Number.isNaN(actual) && Number.isNaN(expected)))
    fail(path, `${JSON.stringify(actual)} != expected ${JSON.stringify(expected)}`);
}

File: tests/helpers/spec-assert.py
"""spec-assert.py — kit-versioned matcher-aware deep-equality comparator (v2.2).

The generated `data` boundary harness (test-author) asserts an actual response
against `fixtures/<case-id>-expected.json` via `spec_assert(actual, expected)`.
Deep equality, EXCEPT that volatile fields in the expected fixture carry a matcher
token instead of a literal (see §5.49). This is the ONLY comparator — no per-project
assertion logic. One file per supported test language; this is the Python one.

Matcher tokens (string values in the expected fixture):

    "<UUID>"            any UUID v4 string
    "<ISO8601>"         any ISO 8601 datetime string
    "<ANY_STRING>"      any non-null string
    "<ANY_NUMBER>"      any finite number (not bool)
    "<UNORDERED>"       any array (presence + shape only)
    "<MATCHES:regex>"   any string matching the pattern (re.search)

Order-insensitive array WITH item checking: wrap the expected array as
    {"<UNORDERED>": [item, item, ...]}
and the actual must be a list that is an order-insensitive deep-equal multiset of
those items.

A matcher token asserts PRESENCE + SHAPE — never "ignore this field". A missing
volatile field still fails (strict key sets). Extra actual keys also fail.

Usage (pytest or any runner):

    from importlib import import_module
    spec_assert = import_module("tests.helpers.spec-assert").spec_assert  # if importable
    # or load directly:
    #   import importlib.util, pathlib
    #   spec = importlib.util.spec_from_file_location("spec_assert", ".../spec-assert.py")
    ...
    spec_assert(actual, expected)          # raises AssertionError with a JSON path on mismatch
"""

from __future__ import annotations

import json
import re
from numbers import Number
from pathlib import Path
from typing import Any

_UUID4 = re.compile(
    r"^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-4[0-9a-fA-F]{3}-[89abAB][0-9a-fA-F]{3}-[0-9a-fA-F]{12}$"
)
_ISO8601 = re.compile(
    r"^\d{4}-\d{2}-\d{2}[Tt ]\d{2}:\d{2}:\d{2}(\.\d+)?([Zz]|[+-]\d{2}:?\d{2})?$"
)
_MATCHES_PREFIX = "<MATCHES:"


class SpecMismatch(AssertionError):
    """Raised when actual does not satisfy the expected fixture (with a JSON path)."""


def load_fixture(path: str | Path) -> Any:
    """Convenience loader for fixtures/<case-id>-*.json."""
    return json.loads(Path(path).read_text(encoding="utf-8"))


def _is_real_number(v: Any) -> bool:
    # bool is a subclass of int in Python — exclude it explicitly.
    return isinstance(v, Number) and not isinstance(v, bool)


def _fail(path: str, msg: str) -> None:
    raise SpecMismatch(f"spec-assert mismatch at {path}: {msg}")


def _match_token(actual: Any, token: str, path: str) -> bool:
    """Return True if `token` is a known matcher and `actual` satisfies it.
    Raises on a known token that is NOT satisfied. Returns False if not a token."""
    if token == "<UUID>":
        if not (isinstance(actual, str) and _UUID4.match(actual)):
            _fail(path, f"expected a UUID v4 string, got {actual!r}")
        return True
    if token == "<ISO8601>":
        if not (isinstance(actual, str) and _ISO8601.match(actual)):
            _fail(path, f"expected an ISO 8601 datetime string, got {actual!r}")
        return True
    if token == "<ANY_STRING>":
        if not isinstance(actual, str):
            _fail(path, f"expected any string, got {type(actual).__name__}")
        return True
    if token == "<ANY_NUMBER>":
        if not _is_real_number(actual):
            _fail(path, f"expected any finite number, got {actual!r}")
        return True
    if token == "<UNORDERED>":
        if not isinstance(actual, list):
            _fail(path, f"expected any array, got {type(actual).__name__}")
        return True
    if token.startswith(_MATCHES_PREFIX) and token.endswith(">"):
        pattern = token[len(_MATCHES_PREFIX):-1]
        if not (isinstance(actual, str) and re.search(pattern, actual)):
            _fail(path, f"expected a string matching /{pattern}/, got {actual!r}")
        return True
    return False


def _unordered_equal(actual: list, items: list, path: str) -> None:
    if not isinstance(actual, list):
        _fail(path, f"expected an array (unordered), got {type(actual).__name__}")
    if len(actual) != len(items):
        _fail(path, f"array length {len(actual)} != expected {len(items)} (unordered)")
    remaining = list(actual)
    for i, exp in enumerate(items):
        found = -1
        for j, act in enumerate(remaining):
            try:
                spec_assert(act, exp, f"{path}[unordered:{i}]")
                found = j
                break
            except SpecMismatch:
                continue
        if found < 0:
            _fail(path, f"no actual element matches expected item #{i}: {exp!r}")
        remaining.pop(found)


def spec_assert(actual: Any, expected: Any, path: str = "$") -> None:
    """Assert `actual` deep-equals `expected`, honoring matcher tokens. Raises
    SpecMismatch (an AssertionError) with a JSON path on the first divergence."""
    # scalar matcher token
    if isinstance(expected, str) and _match_token(actual, expected, path):
        return

    # {"<UNORDERED>": [...]} wrapper → order-insensitive item compare
    if isinstance(expected, dict) and list(expected.keys()) == ["<UNORDERED>"]:
        _unordered_equal(actual, expected["<UNORDERED>"], path)
        return

    if isinstance(expected, dict):
        if not isinstance(actual, dict):
            _fail(path, f"expected object, got {type(actual).__name__}")
        exp_keys, act_keys = set(expected), set(actual)
        if exp_keys != act_keys:
            missing = exp_keys - act_keys
            extra = act_keys - exp_keys
            _fail(path, f"key mismatch (missing={sorted(missing)}, extra={sorted(extra)})")
        for k in expected:
            spec_assert(actual[k], expected[k], f"{path}.{k}")
        return

    if isinstance(expected, list):
        if not isinstance(actual, list):
            _fail(path, f"expected array, got {type(actual).__name__}")
        if len(actual) != len(expected):
            _fail(path, f"array length {len(actual)} != expected {len(expected)}")
        for i, (a, e) in enumerate(zip(actual, expected)):
            spec_assert(a, e, f"{path}[{i}]")
        return

    # scalar literal — strict equality (bool/int kept distinct)
    if type(actual) is not type(expected) and not (
        _is_real_number(actual) and _is_real_number(expected)
    ):
        _fail(path, f"type {type(actual).__name__} != expected {type(expected).__name__}")
    if actual != expected:
        _fail(path, f"{actual!r} != expected {expected!r}")
